### PR TITLE
GasGadget for GAS opcode

### PR DESCRIPTION
**Pull request by scroll-dev**
Implement the GasGadget for opcode GAS (5A). Specs at https://github.com/appliedzkp/zkevm-specs/pull/104.

Other changes:

1. Fix constant gas cost for OpcodeId::EXP

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/param.rs [hunk 1]
```diff
@@ -32,3 +30,4 @@ pub(crate) const STACK_CAPACITY: usize = 1024;
 // counter to u64 as go-ethereum in case transaction size is allowed larger in
 // the future.
 pub(crate) const N_BYTES_PROGRAM_COUNTER: usize = N_BYTES_U64;
+pub(crate) const N_BYTES_GAS: usize = N_BYTES_U64;
```
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/param.rs:33]: Add a doc-comment for this one please.

File: zkevm-circuits/src/evm_circuit/execution/gas.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::test_util::run_test_circuits;
+    use bus_mapping::bytecode;
+
+    fn test_ok() {
```
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit/execution/gas.rs:100]: Can we add a test that verifies a failure on gas deduction too? Not like out of gas, but like a wrong substraction value.

**Review by CPerezz — COMMENTED** [review]
LGTM once https://github.com/appliedzkp/zkevm-specs/pull/104 has been merged.

Also, some small nits.

Great work!!

File: zkevm-circuits/src/evm_circuit/execution/gas.rs (hunk 2 above)
Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/gas.rs:100]: Do you mean a test with an invalid witness?

**Review by ed255 — APPROVED** [review]
LGTM! (Modulo adding a line documenting `N_BYTES_GAS` like Carlos asked)

File: zkevm-circuits/src/evm_circuit/execution/gas.rs (hunk 2 above)
Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/gas.rs:100]: Do you mean, assigning an incorrect `gas_left` for the next step (STOP) and asserting that the circuit cannot be verified?

**Comment by roynalnaruto** [comment]
Not sure why the `rustfmt` check fails. I'm using the same toolchain, and `cargo fmt --all -- --check` gives me no warnings/errors.

**Comment by ed255** [comment]
> Not sure why the `rustfmt` check fails. I'm using the same toolchain, and `cargo fmt --all -- --check` gives me no warnings/errors.

There's a file in `circuit-benchmarks/` that is generated on build time.  If you run `cargo fmt` when this file is not available, it will not complete successfully because the file is not found and the process aborts, skipping some files.   Can you try this?:
```sh
cargo check --all-features # required to generate `circuit-benchmarks/src/bench_params.rs`
cargo fmt --all
```

**Comment by ed255** [comment]
Thanks for the contribution!

**Pull request merged**